server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/SnapshotBasedIndexRecoveryIT.java

```diff
@@ -0,0 +1,611 @@
+                    }
+
+                    return new FilterInputStream(super.readBlob(blobName)) {
+                        @Override
+                        public int read(byte[] b, int off, int len) throws IOException {
+                            int read = super.read(b, off, len);
+                            if (read > 0) {
+                                totalBytesRead.addAndGet(read);
+                            }
+                            return read;
+                        }
+                    };
+                }
+            };
+        }
+    }
+
+    public static class FaultyRepository extends FsRepository {
+        public FaultyRepository(RepositoryMetadata metadata,
+                                Environment environment,
+                                NamedXContentRegistry namedXContentRegistry,
+                                ClusterService clusterService,
+                                BigArrays bigArrays,
+                                RecoverySettings recoverySettings) {
+            super(metadata, environment, namedXContentRegistry, clusterService, bigArrays, recoverySettings);
+        }
+
+        @Override
+        public BlobContainer shardContainer(IndexId indexId, int shardId) {
+            return new FilterBlobContainer(super.shardContainer(indexId, shardId)) {
+                @Override
+                protected BlobContainer wrapChild(BlobContainer child) {
+                    return child;
+                }
+
+                @Override
+                public InputStream readBlob(String blobName) throws IOException {
+                    // Fail only in index files
+                    if (blobName.startsWith("__") == false) {
+                        return super.readBlob(blobName);
+                    }
+
+                    return new FilterInputStream(super.readBlob(blobName)) {
+                        @Override
+                        public int read(byte[] b, int off, int len) throws IOException {
+                            if (randomBoolean()) {
+                                // Return random data
+                                for (int i = 0; i < len; i++) {
+                                    b[off + i] = randomByte();
+                                }
+                                return len;
+                            } else {
+                                if (randomBoolean()) {
+                                    throw new IOException("Unable to read blob " + blobName);
+                                } else {
+                                    // Skip some file chunks
+                                    int read = super.read(b, off, len);
+                                    return read / 2;
+                                }
+                            }
+                        }
+                    };
+                }
+            };
+        }
+    }
+
+    public void testPeerRecoveryUsesSnapshots() throws Exception {
```

Comment:
Could we assert that no file chunks get transferred during this test?

Reply:
I can instrument the TransportService and count the number of `FILE_CHUNK` requests, wdyt?

Reply:
yep sounds good